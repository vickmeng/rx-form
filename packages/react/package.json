{
  "name": "@yzw-rx-form/react",
  "version": "1.2.0",
  "description": "core",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "main": "dist/index.js",
  "private": false,
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c",
    "watch": "rollup -c --watch"
  },
  "author": "laomeng",
  "license": "ISC",
  "peerDependencies": {
    "react": "*",
    "@yzw-rx-form/core": "workspace:*"
  },
  "devDependencies": {
    "react": "*",
    "@yzw-rx-form/core": "workspace:*"
  }
}
