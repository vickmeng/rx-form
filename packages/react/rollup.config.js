import typescript from '@rollup/plugin-typescript';

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'es',
  },
  external: ['react'],
  plugins: [
    typescript({
      tsconfig: './tsconfig.json',
    }),
  ],
};
