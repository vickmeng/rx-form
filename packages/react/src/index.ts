/**
 * components
 */
export { Field } from './items/field';
export { Group } from './items/group';
export { List } from './items/list';
export { Error } from './items/error';

export type { FieldInternalProps, GroupInternalProps, ListInternalProps, ErrorInternalProps } from './types';

export {
  useControlValue,
  useControlDisabled,
  useControlDirty,
  useControlValid,
  useControlErrors,
  useControlControls,
  useControlAsyncErrors,
} from './hooks';
