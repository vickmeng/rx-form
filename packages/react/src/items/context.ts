import React from 'react';
import { GroupControl, ListControl } from '@rx-form/core';

export const ParentFormContext = React.createContext<GroupControl | ListControl<any> | null>(null);
