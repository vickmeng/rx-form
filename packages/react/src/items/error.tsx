import { useContext } from 'react';
import { BaseControl, GroupControl } from '@yzw-rx-form/core';

import { ErrorInternalProps, ErrorProps } from '../types';
import { __throwRxFormReactError, isErrorWithNameProps } from '../utils';
import { useControlDirty, useControlDisabled, useControlErrors, useControlValid } from '../hooks';

import { ParentFormContext } from './context';

export const Error = (props: ErrorProps) => {
  const { children } = props;

  const parentGroup = useContext(ParentFormContext);

  const { name = undefined, control } = isErrorWithNameProps(props)
    ? { name: props.name, control: parentGroup!.get<GroupControl>(props.name) }
    : { control: props.control };

  if (!(control instanceof BaseControl)) {
    __throwRxFormReactError(`Error组件控制器绑定错误，当前name为${name}`);
  }

  const disabled = useControlDisabled(control);
  const dirty = useControlDirty(control);
  const errors = useControlErrors(control);
  const valid = useControlValid(control);

  const childrenProps: ErrorInternalProps = {
    name,
    control,
    errors,
    disabled,
    valid,
    dirty,
  };
  return children(childrenProps);
};
