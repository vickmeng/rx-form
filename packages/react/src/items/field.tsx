import { useContext } from 'react';
import { FieldControl } from '@yzw-rx-form/core';

import { FieldInternalProps, FieldProps } from '../types';
import { isFieldWithNameProps } from '../utils';
import { useControlDirty, useControlDisabled, useControlErrors, useControlValid, useControlValue } from '../hooks';

import { ParentFormContext } from './context';

export function Field<V>(props: FieldProps<V>) {
  const { children } = props;

  const parent = useContext(ParentFormContext);

  /**
   * Two ways can get formControl,from props or Context
   */
  const { name = undefined, control } = isFieldWithNameProps<V>(props)
    ? { name: props.name, control: parent!.get<FieldControl<V>>(props.name) }
    : { control: props.control };

  if (!(control instanceof FieldControl)) {
    throw new Error('props error:Field can only receive FieldControl as control');
  }

  const value = useControlValue<V>(control);
  const disabled = useControlDisabled(control);
  const dirty = useControlDirty(control);
  const valid = useControlValid(control);
  const errors = useControlErrors(control);
  // const asyncErrors = useControlAsyncErrors(control);

  const childrenProps: FieldInternalProps = {
    name,
    value,
    control,
    setValue: control.setValue,
    markAsDirty: control.markAsDirty,
    markAsPristine: control.markAsPristine,
    validateAndUpdateErrors: control.validateAndUpdateErrors,
    errors,
    // asyncErrors,
    disabled,
    valid,
    dirty,
  };

  return children(childrenProps);
}
