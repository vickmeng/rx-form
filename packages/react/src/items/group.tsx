import React, { useContext } from 'react';
import { GroupControl, GroupValue } from '@yzw-rx-form/core';

import { GroupInternalProps, GroupProps } from '../types';
import { isGroupWithNameProps } from '../utils';
import { useControlControls, useControlDisabled } from '../hooks';

import { ParentFormContext } from './context';

export const Group = <V extends GroupValue = any>(props: GroupProps<V>) => {
  const { children } = props;

  const parent = useContext(ParentFormContext);

  const { name = undefined, control } = isGroupWithNameProps(props)
    ? { name: props.name, control: parent!.get<GroupControl<V>>(props.name) }
    : { control: props.control };

  if (!(control instanceof GroupControl)) {
    throw new Error('props error:Group can only receive GroupControl as control');
  }
  const disabled = useControlDisabled(control);
  const controls = useControlControls<GroupControl<V>>(control);

  const childProps: GroupInternalProps<V> = {
    name,
    disabled,
    control,
    childControls: controls,
  };

  return <ParentFormContext.Provider value={control}>{children(childProps)}</ParentFormContext.Provider>;
};
