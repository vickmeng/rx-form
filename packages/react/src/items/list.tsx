import React, { useContext } from 'react';
import { GroupControl, ListControl } from '@rx-form/core';

import { __throwRxFormReactError, isListWithNameProps } from '../utils';
import { ListInternalProps, ListProps } from '../types';
import { useControlControls, useControlDisabled } from '../hooks';

import { ParentFormContext } from './context';

export const List = <V,>(props: ListProps<V>) => {
  const { children } = props;

  const parent = useContext(ParentFormContext);

  const { name = undefined, control } = isListWithNameProps(props)
    ? { name: props.name, control: parent!.get<ListControl<V>>(props.name) }
    : { control: props.control };

  if (!(control instanceof ListControl)) {
    __throwRxFormReactError(`List组件控制器绑定错误，当前name为${name}`);
  }

  const disabled = useControlDisabled(control);
  const controls = useControlControls<ListControl<V>>(control);

  const childProps: ListInternalProps<V> = {
    name,
    disabled,
    control,
    childControls: controls,
  };

  return <ParentFormContext.Provider value={control}>{children(childProps)}</ParentFormContext.Provider>;
};
