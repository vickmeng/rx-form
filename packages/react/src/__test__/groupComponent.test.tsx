import React from 'react';
import { render } from '@testing-library/react';
import { GroupControl } from '@yzw-rx-form/core';

import { Field } from '../items/field';
import { Group } from '../items/group';

describe('<Group/>', () => {
  it('should render <Group/>', () => {
    const groupControl = new GroupControl({
      name: ['TOM'],
      address: ['No.1,Chaowai Street,Chaoyang District,Beijing City'],
    });
    const { asFragment } = render(
      <Group control={groupControl}>
        {(props) => (
          <>
            <Field<string> name="name">{({ value }) => <input value={value} />}</Field>
            <Field<string> name="address">{({ value }) => <input value={value} />}</Field>
          </>
        )}
      </Group>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
