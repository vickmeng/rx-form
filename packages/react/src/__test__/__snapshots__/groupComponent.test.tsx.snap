// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Group/> should render <Group/> 1`] = `
<DocumentFragment>
  <input
    value="TOM"
  />
  <input
    value="No.1,Chaowai Street,Chaoyang District,Beijing City"
  />
</DocumentFragment>
`;
